### ISSUE TITLE
Assigning data to the dataLayer variable

### ISSUE THREAD

**Issue by sevfurneaux** [nystudio107/craft-seomatic#134]
Is it still possible as the previous Craft 2 plugin, to assign the `dataLayer` variable in a twig template?

For example, I have done this in the Craft `_layout.html` file (above where the SEOMatic snippet is rendered):

```
{% set dataLayer = {
  test: 'test data'
} %}
```

However, having viewed the source, the snippet has not detected the `dataLayer` variable data :

`dataLayer = [];`

I've also console logged `dataLayer`, and this is not showing the `test` attribute.

From what I understand, the twig `dataLayer` variable is detected like this in the  **Script Template**  field in the on the **Google Tag Manager Settings** page:

`{{ dataLayerVariableName.value }} = [{% if dataLayer is defined and dataLayer %}{{ dataLayer |json_encode() |raw }}{% endif %}];`

Its worth noting that I'm not using the `{% hook 'seomaticRender' %}` (which was in the Craft 2 wiki documentation) and am relying on the following *Script Render Location* option:

<img width="445" alt="screen shot 2018-06-06 at 17 19 41" src="https://user-images.githubusercontent.com/1056416/41047889-d372f25c-69ad-11e8-91e4-a77a3b707b25.png">

**Comment by sevfurneaux** [comment]
☝️ I am using version 3.0.18.

**Comment by khalwat** [comment]
So the code that it renders is:

```
{% if googleTagManagerId.value is defined and googleTagManagerId.value %}
{{ dataLayerVariableName.value }} = [{% if dataLayer is defined and dataLayer %}{{ dataLayer |json_encode() |raw }}{% endif %}];
(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
'{{ googleTagManagerUrl.value }}?id='+i+dl;f.parentNode.insertBefore(j,f);
})(window,document,'script','{{ dataLayerVariableName.value }}','{{ googleTagManagerId.value }}');
{% endif %}
```
But if you're setting the Twig `dataLayer` variable, and it's not propagating, I think it might be because the template is rendered in a different context (rather than the global context of the page).

**Comment by sevfurneaux** [comment]
The `dataLayer` variable is set prior to the `<body>` tag in the `_layout.html`:

```
  {% set dataLayer = {
    test: 'Test'
  } %}
  <body>
```

To further test, I moved this to `index.html`, before the `{% block content %}`, but no luck with this either.

I wondered, can you show me example of the `dataLayer` twig tag in use? For example, in a template that you have used. Maybe I can replicate this.

**Comment by sevfurneaux** [comment]
Also, the code previously provided for the tag snippet is the same as we have:

```
{% if googleTagManagerId.value is defined and googleTagManagerId.value %}
{{ dataLayerVariableName.value }} = [{% if dataLayer is defined and dataLayer %}{{ dataLayer |json_encode() |raw }}{% endif %}];
(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
'{{ googleTagManagerUrl.value }}?id='+i+dl;f.parentNode.insertBefore(j,f);
})(window,document,'script','{{ dataLayerVariableName.value }}','{{ googleTagManagerId.value }}');
{% endif %}
```

**Comment by khalwat** [comment]
So you were correct that this is not working as intended; the reason ended up being what I mentioned, which is that the script template isn't rendered in the same context as your Craft templates.

Fixed in https://github.com/nystudio107/craft-seomatic/commit/15101e737836c96bc92da0d445bbd6ad5de8d044

This will be released in the next update of SEOmatic. You'll be able to set the `dataLayer` passed in to Google Tag Manager via Twig:

```twig
{% do seomatic.script.get('googleTagManager').dataLayer({
    'woof': 'bark'
}) %}
```
I updated the documentation to reflect this as well.
